lib/active_merchant/billing/gateways/fluidpay.rb

```diff
@@ -0,0 +1,318 @@
+      self.money_format = :cents
+
+      self.homepage_url = 'https://www.fluidpay.com/'
+      self.display_name = 'Fluid Pay'
+
+      ACTIONS = {
+        authorize: 'authorize',
+        capture: 'capture',
+        purchase: 'purchase',
+        refund: 'refund',
+        verify: 'verify',
+        void: 'void'
+      }
+
+      STANDARD_ERROR_CODE_MAPPING = {
+        200 => STANDARD_ERROR_CODE[:card_declined],
+        223 => STANDARD_ERROR_CODE[:expired_card],
+        225 => STANDARD_ERROR_CODE[:invalid_cvc],
+        240 => STANDARD_ERROR_CODE[:call_issuer],
+        250 => STANDARD_ERROR_CODE[:pickup_card],
+        251 => STANDARD_ERROR_CODE[:pickup_card],
+        252 => STANDARD_ERROR_CODE[:pickup_card],
+        253 => STANDARD_ERROR_CODE[:pickup_card],
+        400 => STANDARD_ERROR_CODE[:processing_error],
+        410 => STANDARD_ERROR_CODE[:config_error]
+      }
+
+      def initialize(options={})
+        requires!(options, :api_key)
+        super
+      end
+
+      def purchase(money, payment, options={})
+        requires!(options, :order_id)
+
+        post = {}
+        add_invoice(post, money, options)
+        add_payment(post, payment)
+        add_3ds(post, payment, options)
+        add_address(post, payment, options)
+        add_customer_data(post, options)
+
+        commit(ACTIONS[:purchase], post)
+      end
+
+      def authorize(money, payment, options={})
+        post = {}
+
+        add_invoice(post, money, options)
+        add_payment(post, payment)
+        add_address(post, payment, options)
+        add_customer_data(post, options)
+
+        commit(ACTIONS[:authorize], post)
+      end
+
+      def capture(money, authorization, options={})
+        options[:transaction_id] = authorization
+
+        post = { amount: amount(money) }
+
+        commit(ACTIONS[:capture], post, options)
+      end
+
+      def refund(money, authorization, options={})
+        options[:transaction_id] = authorization
+
+        post = {
+          amount: amount(money),
+        }
+
+        commit(ACTIONS[:refund], post, options)
+      end
+
+      def void(authorization, options={})
+        options[:transaction_id] = authorization
+
+        commit(ACTIONS[:void], {}, options)
+      end
+
+      def verify(credit_card, options={})
+        post = {}
+
+        add_invoice(post, 100, options)
+        add_payment(post, credit_card)
+        add_address(post, credit_card, options)
+        add_customer_data(post, options)
+
+        commit(ACTIONS[:verify], post)
+      end
+
+      def supports_scrubbing?
+        false
+      end
+
+      def scrub(transcript)
+        raise NotImplementedError
+      end
+
+      private
+
+      def add_customer_data(post, options)
+        if options[:email]
+          post[:email_address] = options[:email]
+
+          post[:billing_address] ||= {}
+          post[:billing_address][:email] = options[:email]
```

Comment:
There is kind of a dependency here that requires add_address to be called before add_customer data.

Not sure it functionally matters, but just noticed that add_customer_data will create the billing_address hash, and add_address will return if there isn't one. Maybe a comment is all we need?

Reply:
The early return in `add_address` is looking for a `:billing_address` key in the `options` hash, but it is confusing since that's the same key the Fluidpay API is expecting in the POST body. Since it's setting values in the `post[:billing_address]` hash in multiple places, I'll update to use the `merge` function instead.